osu.Game.Tests/NonVisual/ModDifficultyAdjustOverridingTest.cs

```diff
@@ -0,0 +1,129 @@
+            Assert.AreEqual(baseDifficulty.OverallDifficulty, mod.OverallDifficulty.Value);
+
+            baseDifficulty = createDifficulty(3);
+            mod.ReadFromDifficulty(baseDifficulty);
+
+            Assert.AreEqual(10, mod.ApproachRate.Value);
+            Assert.AreEqual(10, mod.CircleSize.Value);
+            Assert.AreEqual(baseDifficulty.DrainRate, mod.DrainRate.Value);
+            Assert.AreEqual(baseDifficulty.OverallDifficulty, mod.OverallDifficulty.Value);
+        }
+
+        [Test]
+        public void TestOverridenAfterBaseTrackOnRevert()
+        {
+            var baseDifficulty = createDifficulty(5);
+            mod.ReadFromDifficulty(baseDifficulty);
+
+            mod.ApproachRate.Value = 10;
+            Assert.AreEqual(10, mod.ApproachRate.Value);
+            Assert.AreEqual(baseDifficulty.ApproachRate, mod.ApproachRate.Default);
+
+            mod.ApproachRate.SetDefault();
+            Assert.AreEqual(baseDifficulty.ApproachRate, mod.ApproachRate.Value);
+
+            baseDifficulty = createDifficulty(3);
+            mod.ReadFromDifficulty(baseDifficulty);
+
+            Assert.AreEqual(baseDifficulty.ApproachRate, mod.ApproachRate.Value);
+        }
+
+        [Test]
+        public void TestOverridenAfterBaseDontTrackWithoutChange()
+        {
+            mod.ReadFromDifficulty(createDifficulty(6));
+
+            mod.ApproachRate.Value = 3;
+            mod.CircleSize.Value = 3;
+
+            Assert.AreEqual(3, mod.ApproachRate.Value);
+            Assert.AreEqual(3, mod.CircleSize.Value);
+
+            mod.ReadFromDifficulty(createDifficulty(3));
+
+            Assert.AreEqual(3, mod.ApproachRate.Value);
+            Assert.AreEqual(3, mod.CircleSize.Value);
+
+            mod.ReadFromDifficulty(createDifficulty(4));
+
+            Assert.AreEqual(3, mod.ApproachRate.Value);
+            Assert.AreEqual(3, mod.CircleSize.Value);
+        }
+
+        [Test]
+        public void TestOverrideBeforeBase()
+        {
+            mod.ApproachRate.Value = 10;
+            mod.CircleSize.Value = 10;
+            mod.DrainRate.Value = 10;
+            mod.OverallDifficulty.Value = 10;
+
+            mod.ReadFromDifficulty(createDifficulty(3));
+
+            Assert.AreEqual(10, mod.ApproachRate.Value);
+            Assert.AreEqual(10, mod.CircleSize.Value);
+            Assert.AreEqual(10, mod.DrainRate.Value);
+            Assert.AreEqual(10, mod.OverallDifficulty.Value);
+        }
+
+        [Test]
+        public void TestOverridenBeforeBaseTracksBack()
+        {
+            mod.ApproachRate.Value = 6;
+            mod.CircleSize.Value = 6;
+
+            mod.ReadFromDifficulty(createDifficulty(6));
+
+            Assert.AreEqual(6, mod.ApproachRate.Value);
+            Assert.AreEqual(6, mod.CircleSize.Value);
+
+            mod.ReadFromDifficulty(createDifficulty(3));
+
+            Assert.AreEqual(3, mod.ApproachRate.Value);
+            Assert.AreEqual(3, mod.CircleSize.Value);
+        }
```

Comment:
I don't understand why this should work the way it is written. As far as I'm concerned, the way difficulty adjust works should be dead simple:

* If someone has set a value manually, keep that setting forever, until
* a revert to the default value was requested, in which case the base difficulty should be used.

I'm not going to go into the actual implementation of this as it is way too convoluted and hard to understand for what I think it actually needs to be doing. Probably because these tests are enforcing a very convoluted behaviour. This is all made worse by the fact that all of this data is kept in-band instead of using nullable types.

Reply:
This is just handling a case where if a difficulty adjust mod is created with certain values, then `ReadFromDifficulty` is called first time with a difficulty that matches those certain values of the mod, then consider the setting is not overriding anything and should keep tracking base difficulty value. (otherwise, I believe this will make the mod's settings always become "overriden" when jumping to song select with a copied mod or something of the sorts)

Maybe documenting each test and what behaviour is it testing against in a song select/multiplayer context would help in that?

Reply:
Relying on order of execution is dodgy any way you slice it. This is a representation problem. If the value hasn't been manually touched by anyone, it should not be deserialised, copied, etc. It should remain the default.

Reply:
> If the value hasn't been manually touched by anyone, it should not be deserialised, copied, etc.

I'm not sure I understand what you mean by this, what should be preventing "deserializing, copying", because that doesn't sound like it's possible, can you elaborate more?

Reply:
It should be passed around unambiguously as null or similar. None of this "maybe this is the map difficulty, or maybe it's what a user set" stuff.